python/lsst/ctrl/mpexec/cli/script/qgraph.py

```diff
@@ -0,0 +1,156 @@
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import logging
+
+from lsst.daf.butler.cli.cliLog import CliLog
+from ... import CmdLineFwk
+
+_log = logging.getLogger(__name__.partition(".")[2])
+
+
+def qgraph(pipeline=None, log_level=(), qgraph=None, skip_existing=False, save_qgraph=None,
+           save_single_quanta=None, qgraph_dot=None, butler_config=None, input=list(), output=None,
+           output_run=None, extend_run=False, replace_run=False, prune_replaced=None, data_query="",
+           show=None):
+    """Implements the command line interface `pipetask qgraph` subcommand,
+    should only be called by command line tools and unit test code that test
+    this function.
+
+    Parameters
+    ----------
+    pipeline : `pipe.base.Pipeline` or None.
+        The pipline object used to generate a qgraph. If this is not `None`
+        then `qgraph` should be `None`.
+    log_level : `list` of `tuple`
+        per-component logging levels, each item in the list is a tuple
+        (component, level), `component` is a logger name or an empty string
+        or `None` for root logger, `level` is a logging level name, one of
+        CRITICAL, ERROR, WARNING, INFO, DEBUG (case insensitive).
+    qgraph : `str` or `None`
+        Path location for a serialized quantum graph definition as a pickle
+        file. If this option is not None then `pipeline` should be `None`.
+    skip_existing : `bool`
+        If all Quantum outputs already exist in the output RUN collection then
+        that Quantum will be excluded from the QuantumGraph. Will only be used
+        if `extended-run` flag is set.
+    save_qgraph : `str` or `None`
+        Path location for storing a serialized quantum graph definition as a
+        pickle file.
+    save_single_quanta : `str` or `None`
+        Format string of locations for storing individual quantum graph
+        definition (pickle files). The curly brace {} in the input string will
+        be replaced by a quantum number.
+    qgraph_dot : `str` or `None`
+        Path location for storing GraphViz DOT representation of a quantum graph.
+    butler_config : `str`, `dict`, or `lsst.daf.butler.Config`
+        If `str`, `butler_config` is the path location of the gen3
+        butler/registry config file. If `dict`, `butler_config` is key value
+        pairs used to init or update the `lsst.daf.butler.Config` instance. If
+        `Config`, it is the object used to configure a Butler.
```

Comment:
Do we need all these options in CLI?

Reply:
no, and CLI says it must be a file. But at this level (according to downstream documentation), you could pass in any of these things, and CLI could be extended to accept e.g. kv pairs.